base/errorshow.jl

```diff
@@ -823,6 +823,42 @@ function _simplify_include_frames(trace)
     return trace[kept_frames]
 end
 
+# Collapse frames that have the same location (in some cases)
+function _collapse_repeated_frames(trace)
+    kept_frames = trues(length(trace))
+    last_frame = nothing
+    for i in 1:length(trace)
+        frame::StackFrame, _ = trace[i]
+        if last_frame !== nothing
+            if frame.file == last_frame.file && frame.line == last_frame.line
+                #=
+                Handles this case:
+
+                f(g, a; kw...) = error();
+                @inline f(a; kw...) = f(identity, a; kw...);
+                f(1)
+
+                which otherwise ends up as:
+
+                [4] #f#4 <-- useless
+                @ ./REPL[2]:1 [inlined]
+                [5] f(a::Int64)
+                @ Main ./REPL[2]:1
+                =#
+                if startswith(sprint(show, last_frame), "#")
+                   kept_frames[i-1] = false
+                end
+
+                # TODO: Detect more cases that can be collapsed
```

Comment:
I *think* it would be OK to do something like this (I'm being sloppy and not checking corner cases):
```julia
m, last_m = frame.linfo.def::Method, last_frame.linfo.def::Method
params, last_params = Base.unwrap_unionall(m.sig), Base.unwrap_unionall(last_m.sig)
if length(last_params) > length(params)
   issame = true
   for i = 1:length(params)
       issame &= params[i] == last_params[i]
   end
   if issame
        kept_frames[i] = false
    end
end
```

That checks whether `frame` is the same as `last_frame` but with fewer args.

Reply:
> That checks whether frame is the same as last_frame but with fewer args.

I was worried about cases like 

```julia
f(args...) = length(args) == 1 ? f(1, 2) :
             length(args) == 2 ? f() :
             error()
f(1)
```

Here you get successive calls to f with `1, 2, 0` arguments but you probably want to show them all?

Reply:
If the line numbers are all the same does it matter?

Reply:
I don't see how that differs from filling in kwargs. In either case, Julia creates separate methods internally to fill defaults, but that's really an implementation detail. You could imagine that the first line of the generated code would be the equivalent of `if !@isdefined(a)...`, and then there would be only one method.

Reply:
My implementation isn't vulnerable to that. It's the number of arguments in the Method signature it's looking at,  not the number of arguments in the MethodInstance.

Reply:
I see! OK I'll try add that and see how it fares.

Reply:
Seems to work fine. I also added a commit that seems to deal with default keywords as well.